py_modules/db_api.py

```diff
@@ -1,125 +1,281 @@
-
-
-def insert(table, data):
-    cols = ", ".join('"{}"'.format(col) for col in data.keys())
-    vals = ", ".join(":{}".format(col) for col in data.keys())
-    sql = 'INSERT INTO "{0}" ({1}) VALUES ({2})'.format(table, cols, vals)
-    conn.cursor().execute(sql, data)
-    conn.commit()
-    return True
-    """example
-	sqlite_insert(conn, 'stocks', {
-		'created_at': '2016-04-17',
-		'type': 'BUY',
-		'amount': 500,
-		'price': 45.00})
-	"""
-
-
-def update(table, where, data):
+current_configuration = configuration.get_current()
+
+
+class MigrationException(Exception):
+    """
+    This exception represents a generic error in the execution of the code of a migration.
+    """
+    pass
+
+
+class NotAMigrationException(MigrationException):
+    """
+    This exception is triggered any time a Migration class is built over a file that doesn't respect the naming
+    convention for ToooMail migrations (yyyy-MM-dd-n-migrationName.py).
+    """
+    pass
+
+
+class AlreadyMigratedException(MigrationException):
+    """
+    This exception is triggered when the system tries to execute a migration that has already been executed.
+    """
+    pass
+
+
+class DifferentChecksumMigrationException(MigrationException):
+    """
+    This exception is triggered when the system finds a migration which name or content has been modified since its
+    execution.
+    """
     pass
 
 
-def delete(table, where, what):
-    cur = conn.cursor()
-    cur.execute("DELETE FROM " + table + " WHERE " + where + " LIKE %" + what + "%")
-    cur.commit()
+class OldStateMigrationException(MigrationException):
+    """
+    This exception is triggered when the system finds a not executed migration that has a date older than the last
+    executed migration.
+    """
+    pass
+
+
+class DBApi:
+
+    def __init__(self, table=None):
+        self.table = table
+        self.conn = sqlite3.connect(current_configuration.db_location,
+                                    detect_types=sqlite3.PARSE_DECLTYPES | sqlite3.PARSE_COLNAMES)
+
+    def __del__(self):
+        self.conn.close()
+
+    def insert(self, data):
+        columns = ", ".join('"{}"'.format(col) for col in data.keys())
+        values = ", ".join(":{}".format(col) for col in data.keys())
+        sql = 'INSERT INTO "{0}" ({1}) VALUES ({2})'.format(self.table, columns, values)
+        self.conn.cursor().execute(sql, data)
+        self.conn.commit()
+        return True
+
+    def delete(self, where, what):
+        cur = self.conn.cursor()
+        cur.execute('DELETE FROM "{0}" WHERE "{1}" LIKE %"{2}"%'.format(self.table, where, what))
+        self.conn.commit()
+        return
+
+    def get(self, field: object = "", expression: object = "") -> object:
+        cur = self.conn.cursor()
+        cur.execute("SELECT " + field + " FROM " + self.table + " " + expression + "")
+        rows = cur.fetchall()
+        data = []
+        for row in rows:
+            data.append(row)
```

Comment:
Maybe It is redundant, fetchall() already returns a list

Reply:
I have never tried. I testing it on the fly right now.

Reply:
Ok fetchall() returns a result set. Technically, it is a tuple of tuples. Each of the inner tuples represent a row in the table.